File: go/book_store.go
package bookstore

func Cost(books []int) int {
	arr := make([]int, 6)
	for _, book := range books {
		arr[book-1]++
	}

	groups := make([]int, 6)

	for {
		ct := 0
		for i, v := range arr {
			if v > 0 {
				ct++
				arr[i]--
			}
		}
		if ct == 0 {
			break
		}
		groups[ct]++
	}

	for groups[3] > 0 && groups[5] > 0 {
		groups[3]--
		groups[5]--
		groups[4] += 2
	}

	discounts := []float64{0, 0, 0.05, 0.1, 0.2, 0.25}

	s := 0
	for i, v := range groups {
		s += i * v * int(800*(1-discounts[i]))
	}
	return s
}

File: elixir/book_store.ex
defmodule BookStore do
  @typedoc "A book is represented by its number in the 5-book series"
  @type book :: 1 | 2 | 3 | 4 | 5

  @discounts [0, 0, 0.05, 0.1, 0.2, 0.25]

  @doc """
  Calculate lowest price (in cents) for a shopping basket containing books.
  """
  @spec total(basket :: [book]) :: integer
  def total(basket) do
    basket
    |> Enum.frequencies()
    |> Stream.unfold(fn
      mp when map_size(mp) == 0 ->
        nil

      mp ->
        next =
          mp
          |> Map.new(fn {k, v} -> {k, v - 1} end)
          |> Map.reject(fn {_, v} -> v == 0 end)

        {map_size(mp), next}
    end)
    |> Enum.frequencies()
    |> then(fn mp ->
      case min(mp[3] || 0, mp[5] || 0) do
        0 ->
          mp

        n ->
          mp
          |> Map.update!(3, &(&1 - n))
          |> Map.update!(5, &(&1 - n))
          |> Map.update(4, n * 2, &(&1 + n * 2))
      end
    end)
    |> Enum.reduce(0, fn {k, v}, acc ->
      acc + k * v * floor(800 * (1 - Enum.at(@discounts, k)))
    end)
  end
end

